File: src/main/java/org/raku/annotation/fix/GrepFirstFix.java
package org.raku.annotation.fix;

import com.intellij.codeInsight.intention.IntentionAction;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.util.IncorrectOperationException;
import org.raku.psi.RakuMethodCall;
import org.raku.psi.RakuPostfixApplication;
import org.jetbrains.annotations.NotNull;

public class GrepFirstFix implements IntentionAction {
    private final RakuMethodCall myGrepCall;
    private final RakuMethodCall myFirstCall;

    public GrepFirstFix(RakuMethodCall element, RakuMethodCall postfix) {
        myGrepCall = element;
        myFirstCall = postfix;
    }

    @NotNull
    @Override
    public String getText() {
        return "Replace .grep.first with .first";
    }

    @NotNull
    @Override
    public String getFamilyName() {
        return getText();
    }

    @Override
    public boolean isAvailable(@NotNull Project project, Editor editor, PsiFile file) {
        return true;
    }

    @Override
    public void invoke(@NotNull Project project, Editor editor, PsiFile file) throws IncorrectOperationException {
        myGrepCall.setName("first");
        RakuPostfixApplication postfixApp = PsiTreeUtil.getParentOfType(myGrepCall, RakuPostfixApplication.class);
        if (postfixApp != null) {
            PsiElement innerCall = postfixApp.copy();
            RakuPostfixApplication outerCall = PsiTreeUtil.getParentOfType(myFirstCall, RakuPostfixApplication.class);
            if (outerCall != null)
                outerCall.replace(innerCall);
        }
    }

    @Override
    public boolean startInWriteAction() {
        return true;
    }
}

File: src/main/java/org/raku/comma/inspection/fixes/GrepFirstFix.kt
package org.raku.comma.inspection.fixes

import com.intellij.codeInspection.LocalQuickFix
import com.intellij.codeInspection.ProblemDescriptor
import com.intellij.openapi.project.Project
import com.intellij.psi.util.PsiTreeUtil
import org.raku.comma.psi.RakuMethodCall
import org.raku.comma.psi.RakuPostfixApplication

class GrepFirstFix : LocalQuickFix {

    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
        val element = descriptor.psiElement

        val firstCall = (element as RakuPostfixApplication).postfix as RakuMethodCall
        val grepCall  = (element.operand as RakuPostfixApplication).postfix as RakuMethodCall

        grepCall.setName("first")

        val postfixApp = PsiTreeUtil.getParentOfType(grepCall, RakuPostfixApplication::class.java)
        val innerCall = postfixApp?.copy() ?: return
        val outerCall = PsiTreeUtil.getParentOfType(firstCall, RakuPostfixApplication::class.java)

        outerCall?.replace(innerCall)
    }

    override fun getFamilyName(): String { return "Replace .grep.first with .first" }
}